### PR TITLE
Is509 correct errors in z390 implementation of  HLASM built-in functions

### PR DESCRIPTION
**Pull request by jyganci**
Many of the z390 implementations of the HLASM built-in functions have errors. This PR fixes all errors found. The regression test programs added, rt/mlc/TOPR2.MLC and rt/test/TESTOPR2.MLC, test all the functions but do not test error conditions. This has been done outside of Git and will be added to the tests in Git at a future date. In additoin, due to a Git text restricton -- no null (X;00') characters in text -- some tests are not part of the tests at this time, although they have been done outside of Git.

### CODE REVIEW

File: rt/test/TESTOPR2.MLC [hunk 1]
```diff
@@ -0,0 +1,1443 @@
...
+* verify results. In addition, the HLASM Language Reference examples
+* for each built-in function are added as TESTOPR2 tests (there are
+* a few exceptions). Where code tests in TOPR2 completely replace
+* the corresponding MNOTE.AIF tests, the original tests are
+* left in but branched around via AGO.
+*
+*----------------------------------------------------------------------
+*
+* Registers on entry:
+*     R15  emtry point
```
Comment (abekornelis) [inline rt/test/TESTOPR2.MLC]: typo

File: rt/test/TESTOPR2.MLC [hunk 2]
```diff
@@ -0,0 +1,1443 @@
...
+*======================================================================
+         MNOTE '********** Begin C2B tests **********'
+*----------------------------------------------------------------------
+*        HLASM LR examples
+
+&C2B     SETC  C2B('')
+         MNOTE 'C2B('''')=>&C2B<'
+         AIF   (K'&C2B NE 0).C2ERR
+***********************************************************************
+*        TODO how to type the null character?
```
Comment (abekornelis) [inline rt/test/TESTOPR2.MLC]: If there is anything left to do: please complete the action and remove the remark. 
Alternatively, open new issue to cover the todo item and equally remove the remark.

File: rt/test/TESTOPR2.MLC [hunk 3]
```diff
@@ -0,0 +1,1443 @@
...
+*======================================================================
+         MNOTE '********** Begin C2D tests **********'
+*----------------------------------------------------------------------
+*        HLASM LR examples
+
+&C2D     SETC  C2D('')                 at most 4 characters
+         MNOTE 'C2D('''')=&C2D'
+         AIF   ('&C2D' NE '+0').C2ERR
+***********************************************************************
+*        TODO: See JJGOPR3.MLC for next test; it succeeded.
```
Comment (abekornelis) [inline rt/test/TESTOPR2.MLC]: Please remove todo text. Reference to items not in the repo are fine while work is in progress, but should not be needed after delivery.

File: rt/test/TESTOPR2.MLC [hunk 4]
```diff
@@ -0,0 +1,1443 @@
...
+*======================================================================
+         MNOTE '********** Begin C2X tests **********'
+*----------------------------------------------------------------------
+*        HLASM LR examples
+
+&C2X     SETC  C2X('')
+         MNOTE 'C2X('''')=>&C2X<'
+         AIF   (K'&C2X NE 0).C2ERR
+***********************************************************************
+*        TODO: See JJGOPR3.MLC for next test; it succeeded.
```
Comment (abekornelis) [inline rt/test/TESTOPR2.MLC]: and another todo item. please see above.

File: rt/test/TESTOPR2.MLC [hunk 5]
```diff
@@ -0,0 +1,1443 @@
...
+*        Using HLASM the MNOTE displays
+*
+* +K'&DCLSTR=17  &DCLSTR = >''''&&&&&&&&1''''<           
+*
+*        Looks like mz390 just shows the MNOTE with the message
+*        text doubled, not the actual output of the MNOTE.
+*
+*        The following DC statement is the same for both z390 & HLASM.
+***********************************************************************
+         PUSH  PRINT
```
Comment (abekornelis) [inline rt/test/TESTOPR2.MLC:771]: I presume this indicates a shortcoming of z390 MNOTE processing. If so, then an issue should be created to cover the case.

File: rt/test/TESTOPR2.MLC [hunk 6]
```diff
@@ -0,0 +1,1443 @@
...
+         DC    FL1'&K',C'>&DDCLSTR<' 
+         POP   PRINT
+*
+&DCLEN   SETA  DCLEN('&DCLSTR') 
+         MNOTE 'DCLEN(''&&DCLSTR'')=&DCLEN' 
+         AIF   (&DCLEN NE 9).DCLERR
+*----------------------------------------------------------------------
+         MNOTE '********** End   DCLEN tests **********'
+*======================================================================
+         AGO   .SKPDCVO    DCVAL tests now done in rt/mlc/TOPR2.MLC
```
Comment (abekornelis) [inline rt/test/TESTOPR2.MLC:782]: if the tests have been moved, then it is (obviously) okay to document that here, but shouldn't the code be removed?

File: rt/test/TESTOPR2.MLC [hunk 7]
```diff
@@ -0,0 +1,1443 @@
...
+&K       SETA  K'&DCVAL
+&DCVAL2  SETC  '&DCVAL'.'&DCVAL'
+         MNOTE 'DCVAL(''&&&&'')=>&DCVAL2<'
+         AIF   ('&DCVAL2' NE '&&').DCVERR
+*
+&DCVAL   SETC  DCVAL('a''''b')           string is "a'b"
+&K       SETA  K'&DCVAL
+*        MNOTE 'DCVAL(''a''''''''b'')=>&DCVAL<  length = &K'
+         AIF   ('&DCVAL' NE 'a''b').DCVERR
+* AGO .JJG3
```
Comment (abekornelis) [inline rt/test/TESTOPR2.MLC:810]: Could you please either document why this test is being skipped, or just remove the code in its entirety?

File: rt/test/TESTOPR2.MLC [hunk 8]
```diff
@@ -0,0 +1,1443 @@
...
+         AIF   (&ISBIN NE 0).ISERR
+***********************************************************************
+*        TODO: Next test now produces
+*                MNOTE 8,'ISBIN invalid operand value; length is zero'
+*              on z390, as it should. Need to add test to error tests
+* 
+*ISBIN   SETA  ISBIN('')                              error condition 
+*        MNOTE 'ISBIN('''')=&ISBIN' 
+*        AIF   (&ISBIN NE 0).ISERR 
+*----------------------------------------------------------------------
```
Comment (abekornelis) [inline rt/test/TESTOPR2.MLC:892]: Another todo item. Please see comments above on todo items

File: rt/test/TESTOPR2.MLC [hunk 9]
```diff
@@ -0,0 +1,1443 @@
...
+         MNOTE 'ISDEC(''+25'')=&ISDEC' 
+         AIF   (&ISDEC NE 0).ISERR 
+&ISDEC   SETA  ISDEC('2147483648')             too large 
+         MNOTE 'ISDEC(''2147483648'')=&ISDEC' 
+         AIF   (&ISDEC NE 0).ISERR 
+&ISDEC   SETA  ISDEC('00000000005')            too many digits; max 10
+         MNOTE 'ISDEC(''00000000005'')=&ISDEC' 
+         AIF   (&ISDEC NE 0).ISERR 
+***********************************************************************
+*        TODO: Next test now produces
```
Comment (abekornelis) [inline rt/test/TESTOPR2.MLC]: yet another todo item.

File: rt/test/TESTOPR2.MLC [hunk 10]
```diff
@@ -0,0 +1,1443 @@
...
+         MNOTE 'ISHEX(''ab3fCD9F'')=&ISHEX' 
+         AIF   (&ISHEX NE 1).ISERR 
+&ISHEX   SETA  ISHEX('abcdEFGH') 
+         MNOTE 'ISHEX(''abcdEFGH'')=&ISHEX'    non-hexadecimal digits
+         AIF   (&ISHEX NE 0).ISERR 
+&ISHEX   SETA  ISHEX('123456789')              too many chars; 8 max 
+         MNOTE 'ISHEX(''123456789'')=&ISHEX' 
+         AIF   (&ISHEX NE 0).ISERR
+***********************************************************************
+*        TODO: Next test now produces
```
Comment (abekornelis) [inline rt/test/TESTOPR2.MLC]: todo item

File: rt/test/TESTOPR2.MLC [hunk 11]
```diff
@@ -0,0 +1,1443 @@
...
+         MNOTE 'ISSYM(''_Abcd1234'')=&ISSYM' 
+         AIF   (&ISSYM NE 1).ISERR 
+&ISSYM   SETA  ISSYM('##@$_') 
+         MNOTE 'ISSYM(''##@$_'')=&ISSYM' 
+         AIF   (&ISSYM NE 1).ISERR 
+&ISSYM   SETA  ISSYM('1234_Abcd')              invalid initial char
+         MNOTE 'ISSYM(''1234_Abcd'')=&ISSYM' 
+         AIF   (&ISSYM NE 0).ISERR
+***********************************************************************
+*        TODO: Next test now produces
```
Comment (abekornelis) [inline rt/test/TESTOPR2.MLC]: todo item

File: rt/test/TESTOPR2.MLC [hunk 12]
```diff
@@ -0,0 +1,1443 @@
...
+         AIF   (&ISSYM NE 1).ISERR 
+&ISSYM   SETA  ISSYM('AZ090#$_**') 
+         MNOTE 'ISSYM(''AZ090#$_**'')=&ISSYM' 
+         AIF   (&ISSYM NE 0).ISERR 
+.SKPIS ANOP
+*----------------------------------------------------------------------
+         MNOTE '********** End   ISSYM tests **********'
+***********************************************************************
+         MBOX  'Test AND, NOT, OR, XOR'
+*AGO .SKPLOG
```
Comment (abekornelis) [inline rt/test/TESTOPR2.MLC:1017]: is this still needed?

File: rt/test/TESTOPR2.MLC [hunk 13]
```diff
@@ -0,0 +1,1443 @@
...
+&XOR     SETA  (&OP1 XOR &OP2) 
+&BXOR    SETC  A2B(&XOR) 
+         MNOTE '(&OP1 XOR &OP2) = &XOR (B''&BXOR'')' 
+         AIF   (&XOR NE 12).LOERR 
+.SKPLOG  ANOP
+*----------------------------------------------------------------------
+         MNOTE '********** End   AND, NOT,OR, XOR tests **********'
+***********************************************************************
+         MBOX  'Test SIGNED, SYSATTRA, SYSATTRP'
+*AGO .SKPSSS
```
Comment (abekornelis) [inline rt/test/TESTOPR2.MLC:1052]: can be removed?

File: rt/test/TESTOPR2.MLC [hunk 14]
```diff
@@ -0,0 +1,1443 @@
...
+         AIF   ('&SYSATTRA' NE 'GR32').SYERR
+&SYSATTRP SETC  SYSATTRP('SYM')                 see HLASM LR pp339-340
+         MNOTE 'SYSATTRP(''SYM'')=&SYSATTRP'
+         AIF   ('&SYSATTRP' NE 'PGMA').SYERR
+.SKPSSS ANOP
+*----------------------------------------------------------------------
+         MNOTE '********** End   SYSATTRA, SYSATTRP tests **********'
+***********************************************************************
+         MBOX  'Test SLA, SLL, SRA, SRL'
+*AGO .SKPSHFT
```
Comment (abekornelis) [inline rt/test/TESTOPR2.MLC:1088]: can be removed?

File: rt/test/TESTOPR2.MLC [hunk 15]
```diff
@@ -0,0 +1,1443 @@
...
+*        z390 tests
+
+&SLA     SETA  (1 SLA 2)
+         MNOTE '(1 SLA 2) = &SLA'
+         AIF   (&SLA NE 4).SLAERR
+&SLA     SETA  (X'3FFFFFFF' SLA 1)
+         MNOTE '(X''3FFFFFFF'' SLA 1) = &SLA'
+         AIF   (&SLA NE X'7FFFFFFE').SLAERR
+***********************************************************************
+*         TODO: Next 2 tests result in arithmetic overflow
```
Comment (abekornelis) [inline rt/test/TESTOPR2.MLC]: todo item

File: rt/test/TESTOPR2.MLC [hunk 16]
```diff
@@ -0,0 +1,1443 @@
...
+&SRL     SETA  (X'FFFFFFFE' SRL 1)
+&XSRL    SETC  A2X(&SRL)
+         MNOTE '(X''FFFFFFFE'' SLL 1) = &SRL (X''&XSRL'')'
+         AIF   (&SRL NE X'7FFFFFFF').SRLERR
+.SKPSHFT ANOP
+*----------------------------------------------------------------------
+         MNOTE '********** End   SRL tests **********'
+***********************************************************************
+         MBOX  'Test operator types A,E,M,O,S,U'
+*AGO .SKPOT
```
Comment (abekornelis) [inline rt/test/TESTOPR2.MLC:1218]: is this still needed?

File: rt/test/TESTOPR2.MLC (hunk 1 above)
Comment (jyganci) [reply rt/test/TESTOPR2.MLC]: fixed.

File: rt/test/TESTOPR2.MLC (hunk 6 above)
Comment (jyganci) [reply rt/test/TESTOPR2.MLC:782]: I left it in so that one could compare "new" TOPR2.MLC version to "old" TESTOPR2.MLC version. Happy to delete it now if you think it is okay to do so.

File: rt/test/TESTOPR2.MLC (hunk 7 above)
Comment (jyganci) [reply rt/test/TESTOPR2.MLC:810]: This AGO is withing the above enclosing AGO .SKPDCV0. It was skipped originally because I could not get the code to work with all the doubled ampersands and single quotes. The new TOPR2.MLC code handles this case. Left it in for posterity. It will be deleted if we agree that the entire DCVAL section here is deleted.

File: rt/test/TESTOPR2.MLC (hunk 12 above)
Comment (jyganci) [reply rt/test/TESTOPR2.MLC:1017]: I left commented AGOs in so that I could uncomment them when testing TESTOPR2.MLC. Happy to go through and delete them now if you think it is the right thing to do.

File: rt/test/TESTOPR2.MLC (hunk 6 above)
Comment (abekornelis) [reply rt/test/TESTOPR2.MLC:782]: I think it adds more confusion than clarification. If anyone wants to go time travelling, the git repo supports that. No need to keep old stuff around.

File: rt/mlc/TOPR2.MLC [hunk 17]
```diff
@@ -0,0 +1,1779 @@
+***********************************************************************
+* z390 - Mainframe assembler emulator and run-time engine
```
Comment (abekornelis) [inline rt/mlc/TOPR2.MLC:2]: I noticed the running the program creates rt/mlc/TOPR2.TXT
I'm not sure what the best approach is - I think we need to do one of the following:
1. add the file to the .gitignore master file in the root
2. choose an extension that is ignored anyway

Since the file is mechanically produced, I think it should not be included in the repo. Without measures, sooner or later, it will.

File: rt/mlc/TESTASC1.MLC [hunk 18]
```diff
@@ -439,14 +440,20 @@ CVTTOHEX2 RT1   CCE,=VD(CVTTOHEX),=AL8(CVTTOHEX) RPI 1044 DIFF ADDR
...
          RT1   CCE,=A(2147483648),=X'80000000'      TEST SETA OVERFLOW
-&AVAL    SETA  2147483648
-&CHEX    SETC  A2X(&AVAL)                     TEST NEG A2X
-         RT1   CCE,=C'&CHEX',=C'80000000'
-         RT1   CCE,=A(2147483649),=X'80000001'      
-&AVAL    SETA  2147483649
-&CHEX    SETC  A2X(&AVAL)
-         RT1   CCE,=C'&CHEX',=C'80000001'
+* Before SETA; now fails, as it should                      #509
+*&AVAL    SETA  2147483648                                  #509
```
Comment (abekornelis) [inline rt/mlc/TESTASC1.MLC:446]: I understand these tests needed to be commented out. 
Yet it is - in my opinion - still useful to have failing edge cases in the test suite.
Should they go into a separate test program that is expected to fail RC=8? 
If yes, could you please the commented-out code into an issue that covers the work?

File: rt/mlc/TESTDC1.MLC [hunk 19]
```diff
@@ -439,14 +440,20 @@ CVTTOHEX2 RT1   CCE,=VD(CVTTOHEX),=AL8(CVTTOHEX) RPI 1044 DIFF ADDR
...
+* Next test succeeds; will fail when 2147483648 flagged as error #509
          RT1   CCE,=A(2147483648),=X'80000000'      TEST SETA OVERFLOW
-&AVAL    SETA  2147483648
-&CHEX    SETC  A2X(&AVAL)                     TEST NEG A2X
-         RT1   CCE,=C'&CHEX',=C'80000000'
-         RT1   CCE,=A(2147483649),=X'80000001'      
-&AVAL    SETA  2147483649
-&CHEX    SETC  A2X(&AVAL)
-         RT1   CCE,=C'&CHEX',=C'80000001'
+* Before SETA; now fails, as it should                      #509
```
Comment (abekornelis) [inline rt/mlc/TESTDC1.MLC:445]: I understand these tests needed to be commented out.
Yet it is - in my opinion - still useful to have failing edge cases in the test suite.
Should they go into a separate test program that is expected to fail RC=8?
If yes, could you please the commented-out code into an issue that covers the work?

File: rt/mlc/TOPR2.MLC (hunk 17 above)
Comment (abekornelis) [reply rt/mlc/TOPR2.MLC:2]: John changed the file's extension to .TST

File: src/az390.java [hunk 20]
```diff
@@ -9414,20 +9415,24 @@ private void process_dcc_data(){
...
+	if (dcc_len > 0 || dc_len_explicit) {                            // #509
+		dcc_ascii_req = 
+			 (dcc_quote == '\'' 
+				 && (    (tz390.opt_ascii 
+					      && dc_type_sfx != 'E'
+					     )
+					  || dc_type_sfx == 'A'
+					)
+    	     )
+		     | dcc_quote == '"';  //RPI5 and RPI73
```
Comment (abekornelis) [inline src/az390.java:9427]: I am confused by the use of | where || is used a couple of lines above. 
Apparently it works - but is it semantically correct in this location?

File: src/mz390.java [hunk 21]
```diff
@@ -1520,6 +1533,32 @@ private void compile_patterns(){
...
+		 */
+		try {                                                                    // #509			
+			binDigitsPattern = Pattern.compile("[01]+");                         // #509
+		} catch (PatternSyntaxException e) {                                     // #509
+			abort_error(6,"binary digits pattern error - " + e.toString());      // #509
+		}                                                                        // #509
+		try {                                                                    // #509			
+			decDigitsPattern = Pattern.compile("[0-9]+");                        // #509
+		} catch (PatternSyntaxException e) {                                     // #509
+			abort_error(7,"decimal digits pattern error - " + e.toString());     // #509
```
Comment (abekornelis) [inline src/mz390.java:1555]: Just a question: why are you using abort_error? Have you considered log_error?

File: src/mz390.java [hunk 22]
```diff
@@ -7263,16 +7389,88 @@ private void exp_push_sdt(){
...
 				break;	
 			default:  // must be ascii number
-				seta_value = get_int_from_string(setc_value,10); 
-				exp_stk_seta[tot_exp_stk_var-1] = seta_value;
+				if (is32BitBinaryInteger(setc_value, 10)){                           // #509
+					seta_value = get_int_from_string(setc_value,10);                 // #509
+					exp_stk_seta[tot_exp_stk_var-1] = seta_value;                    // #509
+				} else {                                                             // #509
+					/*
+					 * Special case: SDT started as "-2147483648". This
```
Comment (abekornelis) [inline src/mz390.java:7409]: Thanks for the extensive explanation. It really helps!

File: src/mz390.java [hunk 23]
```diff
@@ -12042,61 +12304,104 @@ private void exec_pc_d2a(){
     	 * convert decimal string to int
     	 */
     	check_setc_quotes(1); // RPI 1139
-    	boolean save_opt_allow = tz390.opt_allow; // RPI 1204
-    	tz390.opt_allow = true;           // RPI 1204
-    	seta_value = get_seta_stack_value(-1);
-    	tz390.opt_allow = save_opt_allow; // RPI 1204
 		setc_value1 = get_setc_stack_value();
+		if (setc_value1.length() == 0) {                                     // #509
+			// HLASM LR states this case indicates an error.                 // #509
```
Comment (abekornelis) [inline src/mz390.java]: I encountered the LR abbreviation in many of your comments. First time I saw it, it took me a while before I realized it was not a LR instruction, but a reference to the Language Reference. Have you defined this abbreviation in each program where you use it? Maybe I'm being over-sensitive. But would prefer LangRef or something like that rather than LR. Just a thought - what do you think?

File: src/az390.java (hunk 20 above)
Comment (jyganci) [reply src/az390.java:9427]: I believe that both work. However, the | is not the proper usage. See the StackOverflow article, second answer:

https://stackoverflow.com/questions/1795808/and-and-or-in-if-statements

File: src/mz390.java (hunk 21 above)
Comment (jyganci) [reply src/mz390.java:1555]: Because all the try-catch pattern compiles above the new ones used it.

File: src/mz390.java (hunk 23 above)
Comment (jyganci) [reply src/mz390.java]: I think I used "HLASM Language Reference" for the first occurrence and "HLASM LR" for all remaining occurrences. I can see where someone jumping in the middle might need what you suggested. Will make the updates.

File: src/mz390.java (hunk 22 above)
Comment (jyganci) [reply src/mz390.java:7409]: You are welcome.

File: src/mz390.java (hunk 21 above)
Comment (abekornelis) [reply src/mz390.java:1555]: okay, fair. No problem.

File: rt/mlc/TESTASC1.MLC (hunk 18 above)
Comment (jyganci) [reply rt/mlc/TESTASC1.MLC:446]: A future issue fixing arithmetic expression processing will contain these tests as error tests.

**Review by abekornelis — APPROVED** [review]
As far as I can see all review comments have been addressed. Please go ahead and merge. Don't forget to select Squash Merge, please!
